File: src/gui/Canvas/Viewport.zig
//! Coordinate transforms between world space and pixel space.

const types = @import("types.zig");

const Point = types.Point;
const Vec2 = types.Vec2;
const RenderViewport = types.RenderViewport;

/// World-to-pixel: convert a world-space point to screen-space pixel coordinates.
pub inline fn w2p(pt: Point, vp: RenderViewport) Vec2 {
    const world: Vec2 = @floatFromInt(@as(@Vector(2, i32), pt));
    const pan: Vec2 = .{ vp.pan[0], vp.pan[1] };
    const s: Vec2 = @splat(vp.scale);
    const center: Vec2 = .{ vp.cx, vp.cy };
    return center + (world - pan) * s;
}

/// Pixel-to-world raw: convert pixel coordinates to unsnapped world-space floats.
pub inline fn p2w_raw(pt: Vec2, vp: RenderViewport) Vec2 {
    const center: Vec2 = .{ vp.cx, vp.cy };
    const s: Vec2 = @splat(vp.scale);
    const pan: Vec2 = .{ vp.pan[0], vp.pan[1] };
    return (pt - center) / s + pan;
}

/// Pixel-to-world snapped: convert pixel coordinates to grid-snapped world-space integers.
pub inline fn p2w(pt: Vec2, vp: RenderViewport, snap: f32) Point {
    const center: Vec2 = .{ vp.cx, vp.cy };
    const s: Vec2 = @splat(vp.scale);
    const pan: Vec2 = .{ vp.pan[0], vp.pan[1] };
    const world = (pt - center) / s + pan;
    const gs: f32 = if (snap > 0) snap else 1.0;
    return .{
        @intFromFloat(@round(world[0] / gs) * gs),
        @intFromFloat(@round(world[1] / gs) * gs),
    };
}

File: crates/display/src/canvas/viewport.rs
use egui::Pos2;

use schemify_handler::App;

/// Coordinate transform between world (schematic i32) and pixel (screen f32).
///
/// Transform: `pixel = center + (world - pan) * zoom`
pub struct CanvasViewport {
    /// Center of the canvas area in pixel coordinates.
    pub center: Pos2,
    /// Current zoom factor (from app state).
    pub zoom: f32,
    /// Pan offset in world coordinates.
    pub pan: [f32; 2],
}

impl CanvasViewport {
    /// Build a viewport from the current app state and canvas rect.
    pub fn from_app(app: &App, rect: egui::Rect) -> Self {
        Self {
            center: rect.center(),
            zoom: app.zoom(),
            pan: app.pan(),
        }
    }

    /// Convert world coordinates to pixel position.
    #[inline]
    pub fn world_to_pixel(&self, wx: f32, wy: f32) -> Pos2 {
        Pos2::new(
            self.center.x + (wx - self.pan[0]) * self.zoom,
            self.center.y + (wy - self.pan[1]) * self.zoom,
        )
    }

    /// Convert pixel position to world coordinates (unsnapped f32).
    #[inline]
    pub fn pixel_to_world(&self, px: f32, py: f32) -> [f32; 2] {
        [
            (px - self.center.x) / self.zoom + self.pan[0],
            (py - self.center.y) / self.zoom + self.pan[1],
        ]
    }

    /// Convert pixel position to snapped world coordinates (i32).
    pub fn snap_to_grid(&self, px: f32, py: f32, grid_size: f32) -> [i32; 2] {
        let [wx, wy] = self.pixel_to_world(px, py);
        let gs = if grid_size > 0.0 { grid_size } else { 1.0 };
        [
            (wx / gs).round() as i32 * gs as i32,
            (wy / gs).round() as i32 * gs as i32,
        ]
    }

    /// World-to-pixel for integer coords (convenience).
    #[inline]
    pub fn w2p(&self, wx: i32, wy: i32) -> Pos2 {
        self.world_to_pixel(wx as f32, wy as f32)
    }
}
